File: dynamiclayout.lua
-- DynamicLayout Lua API — generate layout JSON from Lua code.
--
-- Single-file. Zero dependencies outside Lua 5.1+ stdlib.
-- Works in Lua, LuaJIT, embedded (nginx, Redis, WoW, Roblox).
--
-- Example:
--   local dl = require("dynamiclayout")
--   local b = dl.begin("My Page")
--   b:fieldset("Info"):label("Hello from Lua!"):endfieldset()
--   b:button("ok", "OK", "primary", true)
--   print(b:_end())

local M = {}

-- Internal JSON builder (manual — zero deps)
local function escape(s) return (s:gsub('[\\"\n\r\t]', {
    ['\\'] = '\\\\', ['"'] = '\\"', ['\n'] = '\\n', ['\r'] = '\\r', ['\t'] = '\\t'
})) end

local Builder = {}
Builder.__index = Builder

function M.begin(title)
    local b = setmetatable({
        _buf = {},
        _depth = 0,
        _comma = false,
        _counter = 0,
    }, Builder)
    b:_puts('{"ui":{')
    b._depth = 2; b._comma = false
    b:_kv("title", title)
    b:_arr("layout")
    return b
end

-- internal
function Builder:_put(c) table.insert(self._buf, c) end
function Builder:_puts(s) table.insert(self._buf, s) end
function Builder:_indent() self:_put('\n'); for _ = 1, self._depth do self:_puts('  ') end end
function Builder:_cm() if self._comma then self:_put(','); self._comma = false end end
function Builder:_kv(key, val)
    self:_cm(); self:_put(','); self:_indent()
    self:_put('"'); self:_puts(key); self:_puts('": "'); self:_puts(escape(val)); self:_put('"')
    self._comma = true
end
function Builder:_kvint(key, val)
    self:_cm(); self:_put(','); self:_indent()
    self:_put('"'); self:_puts(key); self:_puts('": '); self:_puts(tostring(val))
    self._comma = true
end
function Builder:_kvbool(key)
    self:_cm(); self:_put(','); self:_indent()
    self:_put('"'); self:_puts(key); self:_puts('": true')
    self._comma = true
end
function Builder:_obj()
    self:_cm(); self:_put(','); self:_indent()
    self:_put('{'); self._depth = self._depth + 1; self._comma = false
end
function Builder:_co()
    self._depth = self._depth - 1; self:_indent(); self:_put('}'); self._comma = true
end
function Builder:_arr(key)
    self:_cm(); self:_put(','); self:_indent()
    self:_put('"'); self:_puts(key); self:_puts('": [')
    self._depth = self._depth + 1; self._comma = false
end
function Builder:_ca()
    self._depth = self._depth - 1; self:_indent(); self:_put(']'); self._comma = true
end
function Builder:_key(prefix)
    self._counter = self._counter + 1
    return prefix .. '_' .. self._counter
end

-- Containers
function Builder:fieldset(title)
    self:_obj(); self:_kv("type", "FIELDSET"); self:_kv("key", self:_key("fs")); self:_kv("title", title); self:_arr("content")
    return self
end
function Builder:endfieldset() self:_ca(); self:_co(); return self end
function Builder:row() self:_obj(); self:_kv("type", "ROW"); self:_kv("key", self:_key("r")); self:_arr("content"); return self end
function Builder:endrow() self:_ca(); self:_co(); return self end

-- Display
function Builder:label(text) self:_obj(); self:_kv("type", "LABEL"); self:_kv("key", self:_key("l")); self:_kv("label", text); self:_co(); return self end
function Builder:alert(message, color) color = color or "info"
    self:_obj(); self:_kv("type", "ALERT"); self:_kv("key", self:_key("a")); self:_kv("message", message); self:_kv("color", color); self:_co(); return self
end
function Builder:badge(title, color) color = color or "primary"
    self:_obj(); self:_kv("type", "BADGE"); self:_kv("key", self:_key("bd")); self:_kv("title", title); self:_kv("color", color); self:_co(); return self
end
function Builder:spacer(px) px = px or 20
    self:_obj(); self:_kv("type", "SPACER"); self:_kv("key", self:_key("sp")); self:_kvint("width", px); self:_co(); return self
end

-- Inputs
function Builder:input(id, label, required, dataType)
    self:_obj(); self:_kv("type", "INPUT"); self:_kv("key", self:_key("i")); self:_kv("id", id); self:_kv("label", label or "")
    if required then self:_kvbool("required") end
    if dataType and dataType ~= "STRING" then self:_kv("dataType", dataType) end
    self:_co(); return self
end
function Builder:textarea(id, label, rows)
    self:_obj(); self:_kv("type", "TEXTAREA"); self:_kv("key", self:_key("ta")); self:_kv("id", id); self:_kv("label", label or "")
    if rows then self:_kvint("rows", rows) end
    self:_co(); return self
end
function Builder:checkbox(id, label)
    self:_obj(); self:_kv("type", "CHECKBOX"); self:_kv("key", self:_key("cb")); self:_kv("id", id); self:_kv("label", label or ""); self:_co(); return self
end
function Builder:select(id, label, pairs)
    self:_obj(); self:_kv("type", "SELECT"); self:_kv("key", self:_key("s")); self:_kv("id", id); self:_kv("label", label)
    self:_arr("values")
    for _, p in ipairs(pairs) do self:_obj(); self:_kv("id", p[1]); self:_kv("displayName", p[2]); self:_co() end
    self:_ca(); self:_co(); return self
end
function Builder:rating(id, label)
    self:_obj(); self:_kv("type", "RATING"); self:_kv("key", self:_key("rt")); self:_kv("id", id); self:_kv("label", label or ""); self:_co(); return self
end

-- Actions
function Builder:button(id, title, color, isDefault)
    self:_obj(); self:_kv("type", "BUTTON"); self:_kv("key", self:_key("btn")); self:_kv("id", id); self:_kv("title", title); self:_kv("color", color or "primary")
    if isDefault then self:_kvbool("default") end
    self:_co(); return self
end

-- Shortcuts
function Builder:section(title, text) self:fieldset(title):label(text):endfieldset(); return self end
function Builder:feedbackform()
    self:fieldset("Your Details")
        :input("name", "Name", true)
        :input("email", "Email", true)
        :textarea("message", "Message", 5)
        :endfieldset()
        :button("send", "Send", "primary", true)
        :button("cancel", "Cancel", "secondary")
    return self
end

-- Output
function Builder:_end()
    self:_ca() -- layout
    self._comma = false; self:_put(',')
    self:_puts('"translations":{},')
    self:_puts('"userAccess":{"cancel":true}')
    self:_puts('}}')
    self:_put('\0')
    return table.concat(self._buf):sub(1, -2) -- strip null terminator
end

return M

File: DynamicLayout.fs
// DynamicLayout F# API — generate layout JSON from F# code.
// Single-file. Zero dependencies. Works on .NET 6+.
// Example:
//   open DynamicLayout
//   let b = Layout.begin "My Page"
//   b.fieldset("Info").label("Hello from F#!").endFieldset()
//   b.button("ok", "OK", "primary", true) |> ignore
//   printfn "%s" (b.end())

module DynamicLayout

open System.Text

type Builder() =
    let sb = StringBuilder()
    let mutable depth = 0
    let mutable comma = false
    let mutable cnt = 0

    member private _.putc(c: char) = sb.Append(c) |> ignore
    member private _.puts(s: string) = sb.Append(s) |> ignore
    member private b.indent() = b.putc('\n'); for _ in 1..depth do b.puts("  ")
    member private b.cm() = if comma then b.putc(','); comma <- false
    member private b.esc(s: string) =
        b.putc('"')
        for c in s do match c with '"' -> b.puts("\\\"") | '\\' -> b.puts("\\\\") | '\n' -> b.puts("\\n") | _ -> b.putc(c)
        b.putc('"')
    member private b.kv(k, v) = b.cm(); b.putc(','); b.indent(); b.esc(k); b.puts(": "); b.esc(v); comma <- true
    member private b.kvint(k, v) = b.cm(); b.putc(','); b.indent(); b.esc(k); b.puts(sprintf ": %d" v); comma <- true
    member private b.kvbool(k) = b.cm(); b.putc(','); b.indent(); b.esc(k); b.puts(": true"); comma <- true
    member private b.obj() = b.cm(); b.putc(','); b.indent(); b.putc('{'); depth <- depth + 1; comma <- false
    member private b.co() = depth <- depth - 1; b.indent(); b.putc('}'); comma <- true
    member private b.arr(k) = b.cm(); b.putc(','); b.indent(); b.esc(k); b.puts(": ["); depth <- depth + 1; comma <- false
    member private b.ca() = depth <- depth - 1; b.indent(); b.putc(']'); comma <- true
    member private b.key(p) = cnt <- cnt + 1; sprintf "%s_%d" p cnt

    member b.fieldset(title) = b.obj(); b.kv("type", "FIELDSET"); b.kv("key", b.key("fs")); b.kv("title", title); b.arr("content"); b
    member b.endFieldset() = b.ca(); b.co(); b
    member b.row() = b.obj(); b.kv("type", "ROW"); b.kv("key", b.key("r")); b.arr("content"); b
    member b.endRow() = b.ca(); b.co(); b
    member b.label(text) = b.obj(); b.kv("type", "LABEL"); b.kv("key", b.key("l")); b.kv("label", text); b.co(); b
    member b.alert(msg, ?color) = let c = defaultArg color "info" in b.obj(); b.kv("type", "ALERT"); b.kv("key", b.key("a")); b.kv("message", msg); b.kv("color", c); b.co(); b
    member b.badge(title, ?color) = let c = defaultArg color "primary" in b.obj(); b.kv("type", "BADGE"); b.kv("key", b.key("bd")); b.kv("title", title); b.kv("color", c); b.co(); b
    member b.spacer(?px) = let p = defaultArg px 20 in b.obj(); b.kv("type", "SPACER"); b.kv("key", b.key("sp")); b.kvint("width", p); b.co(); b
    member b.input(id, label, ?required) = b.obj(); b.kv("type", "INPUT"); b.kv("key", b.key("i")); b.kv("id", id); b.kv("label", label); if defaultArg required false then b.kvbool("required"); b.co(); b
    member b.checkbox(id, ?label) = let l = defaultArg label "" in b.obj(); b.kv("type", "CHECKBOX"); b.kv("key", b.key("cb")); b.kv("id", id); b.kv("label", l); b.co(); b
    member b.textarea(id, label, ?rows) = let r = defaultArg rows 3 in b.obj(); b.kv("type", "TEXTAREA"); b.kv("key", b.key("ta")); b.kv("id", id); b.kv("label", label); if r > 0 then b.kvint("rows", r); b.co(); b
    member b.rating(id, ?label) = let l = defaultArg label "" in b.obj(); b.kv("type", "RATING"); b.kv("key", b.key("rt")); b.kv("id", id); b.kv("label", l); b.co(); b
    member b.button(id, title, color, ?isDefault) = b.obj(); b.kv("type", "BUTTON"); b.kv("key", b.key("btn")); b.kv("id", id); b.kv("title", title); b.kv("color", color); if defaultArg isDefault false then b.kvbool("default"); b.co(); b
    member b.section(title, text) = b.fieldset(title).label(text).endFieldset()
    member b.end() = b.ca(); comma <- false; b.putc(','); b.puts("\"translations\":{},"); b.puts("\"userAccess\":{\"cancel\":true}"); b.puts("}}"); sb.ToString()
    static member begin title = let b = Builder() in b.puts("{\"ui\":{"); b.depth <- 2; b.comma <- false; b.kv("title", title); b.arr("layout"); b
